[Cocoa] Reduce telemetry reports for virtual memory routines

brentfulgham opened the pull request:
#### d3c7b8f39ab3b9c551adbaec0bc039834573b9c6
<pre>
[Cocoa] Reduce telemetry reports for virtual memory routines
<a href="https://bugs.webkit.org/show_bug.cgi?id=256736">https://bugs.webkit.org/show_bug.cgi?id=256736</a>
&lt;rdar://105161083&gt;

Reviewed by Per Arne Vollan.

Add additional rules for VM handling features to reduce sandbox telemetry.

* Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in:
* Source/WebKit/NetworkProcess/mac/com.apple.WebKit.NetworkProcess.sb.in:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.GPU.sb.in:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Networking.sb.in:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.adattributiond.sb.in:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.webpushd.sb.in:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
* Source/WebKit/webpushd/mac/com.apple.WebKit.webpushd.mac.sb.in:

Canonical link: <a href="https://commits.webkit.org/264130@main">https://commits.webkit.org/264130@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ec95ec5713b58b3994a5fc62f9230c0dd528ba03

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6812 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7032 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7216 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8403 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7065 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6812 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8211 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6976 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9956 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6931 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7531 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6250 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8498 "Built successfully") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/6922 "webkitpy-tests (failure)") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/4955 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6168 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13970 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/6683 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6250 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9028 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6742 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5521 "Passed tests") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/6774 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6124 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/6083 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1609 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10303 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6502 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in:
```diff
@@ -226,7 +226,8 @@
         "kern.ostype"
-        "kern.version")
+        "kern.version"
+        "vm.malloc_ranges") ;; <rdar://problem/105161083>
```
I think we only need to allow this on startup, which means we can create a rule with the WebContentProcessLaunched state variable.

brentfulgham replied:
Will do!

pvollan commented on Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in:
```diff
@@ -1090,6 +1091,7 @@
                 mach_port_set_attributes
                 mach_vm_copy
                 mach_vm_map_external
+                (when (defined? 'mach_vm_range_create) mach_vm_range_create) ;; <rdar://105161083>
```
Ditto.

pvollan replied:
These comments only apply to the WebContent process profile :)

pvollan commented on Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
```diff
@@ -1152,9 +1152,11 @@
 (with-filter (require-not (state-flag "WebContentProcessLaunched"))
-    (allow syscall-unix (syscall-unix-only-in-use-during-launch)))
+    (allow syscall-unix (syscall-unix-only-in-use-during-launch))
+    (allow sysctl-read (sysctl-name "vm.malloc_ranges"))) ;; <rdar://problem/105161083>
```
Thanks! In order to sort the sandbox by operation, I think it would be good to put this close to the other sysctl-reads.

brentfulgham replied:
Will change.

pvollan commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -2055,7 +2055,10 @@
 (with-filter (require-not (state-flag "WebContentProcessLaunched"))
-    (allow syscall-unix (syscall-number SYS_quotactl)))
+    (allow syscall-unix (syscall-number SYS_quotactl))
+    (allow sysctl-read (sysctl-name "vm.malloc_ranges"))) ;; <rdar://problem/105161083>
```
Ditto.

pvollan commented on Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
```diff
@@ -1404,6 +1411,7 @@
         mach_port_set_attributes
         mach_vm_copy
         mach_vm_map_external
+        (when (defined? 'mach_vm_range_create) mach_vm_range_create) ;; <rdar://105161083>
```
I think also this sys call could only be allowed before the process has finished launching.

pvollan commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -2148,6 +2155,7 @@
     mach_port_set_attributes
     mach_vm_copy
     mach_vm_map_external
+    (when (defined? 'mach_vm_range_create) mach_vm_range_create) ;; <rdar://105161083>
```
I think also this sys call could only be allowed before the process has finished launching.

brentfulgham replied:
I don't see a set of pre-launch mig routines in this sandbox, like there is for iOS. What's the best place to put such a conditional rule?

pvollan replied:
Ah, good point! I think it could be put beneath the standard list of mig routines.

brentfulgham replied:
I took a stab at making that work. My local build works properly on macOS, so I think it is correct. Thankfully, iOS was just adding to your existing "launch" collection of routines.

pvollan commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -2193,7 +2205,23 @@
 
-            (allow mach-message-send (kernel-mig-routines-common)))))
+            (allow mach-message-send
+                (kernel-mig-routine-only-in-use-during-launch)
```
I think it may be more compact to not allow kernel-mig-routine-only-in-use-during-launch here.

pvollan commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -2193,7 +2205,23 @@
+                (deny mach-message-send
+                    (with telemetry)
+                    (with message "kernel mig routine used after launch")
+                    (kernel-mig-routine-only-in-use-during-launch)))
```
Then you can remove this block.

pvollan commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -2193,7 +2205,23 @@
+                    (with message "kernel mig routine used after launch")
+                    (kernel-mig-routine-only-in-use-during-launch)))
+#else
+            (allow mach-message-send (kernel-mig-routine-only-in-use-during-launc))
```
Small typo at the end.

brentfulgham replied:
Oh, no!

pvollan reviewed (approved):
Looks good! R=me.

brentfulgham committed:
[Cocoa] Reduce telemetry reports for virtual memory routines
https://bugs.webkit.org/show_bug.cgi?id=256736
<rdar://105161083>

Reviewed by Per Arne Vollan.

Add additional rules for VM handling features to reduce sandbox telemetry.

* Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in:
* Source/WebKit/NetworkProcess/mac/com.apple.WebKit.NetworkProcess.sb.in:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.GPU.sb.in:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Networking.sb.in:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.adattributiond.sb.in:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.webpushd.sb.in:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
* Source/WebKit/webpushd/mac/com.apple.WebKit.webpushd.mac.sb.in:

Canonical link: https://commits.webkit.org/264130@main
